src/modules/iotjs_module_tls.c

```diff
@@ -0,0 +1,162 @@
+/* Copyright 2018-present Samsung Electronics Co., Ltd. and other contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "iotjs_module_tls.h"
+
+#include "stdarg.h"
+
+
+IOTJS_DEFINE_NATIVE_HANDLE_INFO_THIS_MODULE(tls);
+
+static void iotjs_tls_destroy(iotjs_tls_t* tls_data) {
+  mbedtls_net_free(&tls_data->server_fd);
+  mbedtls_x509_crt_free(&tls_data->cacert);
+  mbedtls_ssl_free(&tls_data->ssl);
+  mbedtls_ssl_config_free(&tls_data->conf);
+  mbedtls_ctr_drbg_free(&tls_data->ctr_drbg);
+  mbedtls_entropy_free(&tls_data->entropy);
+
+  IOTJS_RELEASE(tls_data);
+}
+
+static iotjs_tls_t* tls_create(const jerry_value_t jobject) {
+  iotjs_tls_t* tls_data = IOTJS_ALLOC(iotjs_tls_t);
+
+  tls_data->jobject = jobject;
+  jerry_set_object_native_pointer(jobject, tls_data, &this_module_native_info);
+
+  mbedtls_net_init(&tls_data->server_fd);
+  mbedtls_ssl_init(&tls_data->ssl);
+  mbedtls_ssl_config_init(&tls_data->conf);
+  mbedtls_x509_crt_init(&tls_data->cacert);
+  mbedtls_ctr_drbg_init(&tls_data->ctr_drbg);
+  mbedtls_entropy_init(&tls_data->entropy);
+
+  return tls_data;
+}
+
+
+JS_FUNCTION(Write) {
+  DJS_CHECK_ARGS(1, string);
+  JS_DECLARE_THIS_PTR(tls, tls_data);
+  iotjs_string_t str = JS_GET_ARG(0, string);
+  int ret = 0;
+  while ((ret = mbedtls_ssl_write(&tls_data->ssl,
+                                  (unsigned char*)(iotjs_string_data(&str)),
+                                  iotjs_string_size(&str))) <= 0) {
+    if (ret != MBEDTLS_ERR_SSL_WANT_READ && ret != MBEDTLS_ERR_SSL_WANT_WRITE) {
+      iotjs_string_destroy(&str);
+      return JS_CREATE_ERROR(COMMON, "write error");
+    }
+  }
+  iotjs_string_destroy(&str);
+
+  return jerry_create_number(ret);
+}
+
+static jerry_value_t tls_connect_helper(iotjs_string_t* h, iotjs_string_t* p,
+                                        iotjs_string_t* hn, char* format, ...) {
+  va_list arg_list;
+  va_start(arg_list, format);
+  char* buf;
+  buf = iotjs_buffer_allocate(sizeof(char) * 256);
+  vsnprintf(buf, sizeof(char) * 256, format, arg_list);
+  va_end(arg_list);
+  jerry_value_t ret_val = jerry_create_string((jerry_char_t*)buf);
+  iotjs_string_destroy(h);
+  iotjs_string_destroy(p);
+  iotjs_string_destroy(hn);
+  iotjs_buffer_release(buf);
+
+  return ret_val;
```

Comment:
Please use `JS_CREATE_ERROR(TYPE, message)` to throw an error.

Reply:
You are right! Thank you.